### PR TITLE
[WIP] Configurable serial ports

### PR DESCRIPTION
**Pull request by bkueng**
This adds a script to generate serial port parameters for every exposed UART port that a board has, and generates the necessary ROMFS startup logic to start the configured driver/service.
The UART port mappings are defined per-board in the cmake board config file.
Implemented pretty much as described in https://github.com/PX4/Firmware/issues/8302#issuecomment-388731631.

QGC will only show params for UART ports that the board actually has. It looks like this:
qgc_serial_config
qgc_serial_config_selection

The RC script will show comprehensible bootup outputs:
```
Starting Main GPS on /dev/ttyS3
Starting MAVLink on /dev/ttyS1
Starting FastRTPS on /dev/ttyS2
``` 

### Drawbacks
- There are mavlink parameters for all ports (e.g. for the mavlink mode)
- Commands/drivers not enabled on a board will still show up in the param dropdown list (mostly a limitation of the param metadata file). It is mainly problematic in 2 cases:
  - fmu-v2 target that has disabled many drivers. Updating the bootloader and flashing v3 will fix this.
  - RTPS has still a separate build target. I think we can now enable it in the default builds.

### TODO
- remove GPS_DEFAULT_UART_PORT
- remove `SENS_EN_*` params for existing UART sensor drivers.
- Remove SYS_COMPANION and update the rcS script (I want to do that after https://github.com/PX4/Firmware/pull/9487 to avoid conflicts) @mcsauder fyi
- add CLI support for `p:<param_name>` where needed
- add the serial port mapping to the remaining boards

Fixes https://github.com/PX4/Firmware/issues/8302

### CODE REVIEW

**Comment by dagar** [comment]
Overall this is looking like a great improvement that's badly needed.

**Driver Configuration per Instance**
Where this starts to fall apart is needing to maintain per module configuration parameters per serial port. How do we grow that for additional driver configuration per serial port? We're already missing some of this and more will inevitably be needed.

What if we did the main configuration at the serial port level (CONFIG and likely BAUD) and had the specific module/driver retain ownership of any additional configuration parameters per instance slot?


**Parameters**
Instead of more `p:<param_name>` parameter magic could these be evaluated in back quotes and passed in?

**Board Configuration**
I like having more of the configuration at the cmake level, let's just be sure to understand the implications with respect to the PX4 level board support. There are also situations like px4fmu-v3 where the TELEM and GPS labels aren't consistent.

**Non-NuttX**
I know we're already on the same page about this, but for the sake of discussion I have to ask. What are we going to do about this? I think having completely different init systems is a much bigger problem than people realize. There are quite a few subtle problems with the current init scripts and most development/testing ends up missing things that are primarily hit by users unable to investigate.

**Comment by LorenzMeier** [comment]
This is awesome!

**Comment by bkueng** [comment]
@dagar 
> What if we did the main configuration at the serial port level (CONFIG and likely BAUD) and had the specific module/driver retain ownership of any additional configuration parameters per instance slot?

How would you assign an instance to a port then? Via CONFIG = 'Rangefinder instance 1' enumerating up to N?

> Instead of more p:<param_name> parameter magic could these be evaluated in back quotes and passed in?

That was my first intention but backquotes require a tmpfs, because the output is stored to a temporary file.

> I like having more of the configuration at the cmake level

I actually don't, because it makes testing individual scripts more complicated, as you need to figure out what information is passed to the script, in which format and where it comes from (passing structured data around in cmake is also error prone and cmake has very peculiar handling of strings/variables). Having per-board configuration files, e.g. in JSON/YAML would IMO reduce build system complexity.

**Comment by dagar** [comment]
> > What if we did the main configuration at the serial port level (CONFIG and likely BAUD) and had the specific module/driver retain ownership of any additional configuration parameters per instance slot?
> 
> How would you assign an instance to a port then? Via CONFIG = 'Rangefinder instance 1' enumerating up to N?

Yes that's one option. You assign each serial port to a particular instance of something and then under that module have as much additional configuration (per instance) as needed. The downside is that we'd either need to be quite careful to always keep the numbering consistent, or make it explicit in some way.

A short term compromise could be to make sure we only mark the relevant params active and keep it small until we can think of something scalable. For example dropping mavlink forwarding and possibly even the data rate. We can get close enough based on the mode and port.

For serial driver/module instances there are already many cases where more parameters per instance are needed (Mavlink, GPS, distance, Iridium, etc), but we're also going to need this general idea in other areas.





> > I like having more of the configuration at the cmake level
> 
> I actually don't, because it makes testing individual scripts more complicated, as you need to figure out what information is passed to the script, in which format and where it comes from (passing structured data around in cmake is also error prone and cmake has very peculiar handling of strings/variables). Having per-board configuration files, e.g. in JSON/YAML would IMO reduce build system complexity.

Well, perhaps I should have said I like it as an alternative to the current mix of hard coded things scattered throughout the code base. I'm still very interested in the board manifest idea (https://github.com/PX4/Firmware/issues/8812).

Dealing with strings and variables in cmake is pretty horrible, and in fact a lot of "modern cmake" is really centered around setting properties on targets (public, private, interface) and sharing them via linking relationship rather than messing with strings. I'm hoping to get us there eventually.

File: Tools/serial/generate_config.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+# with QGC
+serial_commands = {
+    0: ["Disabled", ""],
+
+    # MAVLink & RTPS
+    1: ["MAVLink", "mavlink start -d ${SER_DEV} ${MAV_ARGS} -x"],
+
+    2: ["MAVLink over Syslink", '''
+syslink start
+mavlink start -d /dev/bridge0 ${MAV_ARGS}
```
Comment (dagar) [inline Tools/serial/generate_config.py]: Does this one need to be configurable or can it be done automatically on a crazyflie?

**Comment by dagar** [comment]
Without ignoring any of the specific comments I made regarding the details of this PR, the broader concern I have with this is that it's just not very modular and ties a number of disparate components and devices together from the entire system centrally and user facing. The set of drivers and modules in any particular build varies substantially. Some of this is fundamental (specific hardware, flash limits, platform compatibility), and some unintentional (lack of config consistency on our part). 

image

**Questions/Concerns**
 - how is a user going to know that their selection isn't actually possible? The parameter change will be taken and they effectively won't get any error (most users don't even know the serial console exists).
 - how do we handle the variability between boards without tying every possible configuration together?
 - how do we maintain an up to date list of all the possible devices?
 - 3rd party adoption/extension
     - how would a company, researcher, or individual extend this cleanly? If they hack it into these jinja scripts or another part of init it's likely to be a future merge conflict. This might seem trivial, but it's extremely common and turns into one more reason it's not worth keeping up with PX4 development and ever potentially contributing in any capacity.
 - what are we going to do on any posix on qurt platform where none of these parameter will even be evaluated? There's really no reason most of these boards should be treated any differently.

These are some of the reasons I wanted to consider complete configuration at the module level (https://github.com/PX4/Firmware/issues/8302#issuecomment-389016299). The main downside is that we lose the inherent port deconfliction. The upside is that it solves all of the same problems as this PR as well as the additional ones I mentioned above. It also has the side benefit of alleviating a great deal of the init script pain and platform fragmentation.

I'm torn between what is still a fantastic improvement over what we currently have, and the possibility of introducing something that makes it difficult if not impossible to ever address the remaining issues. Thoughts? Dev call? @bkueng @LorenzMeier @thomasgubler

**Comment by LorenzMeier** [comment]
@bkueng Can you rebase?

**Comment by LorenzMeier** [comment]
@dagar Most of these issues can be consistently addressed. And even more importantly: In absence of something better, we should improve the status quo and then disrupt that again.

At the end of the day we have to be realistic about our available development bandwidth.

**Comment by dagar** [comment]
Take mavlink as an example. We have X serial ports and M mavlink instances to configure. Serial ports may or may not be mavlink instances, and mavlink instances may or may not be on serial ports. 

Configuring the function (mavlink, gps, etc) of each serial port directly initially seems like a good idea (port and baudrate), but quickly falls apart when you take the next logical step of wanting to configure additional properties for each of those instances. 

This PR introduces mavlink configuration parameters per serial port (SER_X_MAV_MDE, SER_X_MAV_R, SER_X_MAV_FWD, etc). This places the burden of configuration of every type (mavlink, gps, lidar, frsky, etc) on every serial port. After all of that we're still unable to configure properties on an instance that doesn't have a corresponding serial port (mavlink networking, iridium, USB, etc). There are already quite a few examples of this in the system, and I see the need for many more in the future.

Doesn't that seem like the wrong mapping fundamentally?

**Comment by LorenzMeier** [comment]
Just discussed this with @bkueng - I think this is the way to go. To simplify the meta data download: I'd like to upload per-board, per-version XML files (@DonLakeFlyer).

Example: ```http://px4.aws.com/meta/<product_id>/<git_hash>.xml``` where `product_id` and `git_hash` come from the MAVLink AUTOPILOT_VERSION packet (which is already in QGC).

**Comment by DonLakeFlyer** [comment]
> To simplify the meta data download: I'd like to upload per-board, per-version XML files

@LorenzMeier You're talking about parameter meta data right?

**Comment by LorenzMeier** [comment]
Yes. So that accurate metadata is easy to grab for QGC.

**Comment by DonLakeFlyer** [comment]
Slight problem is that once the vehicle is connected the parameter meta data is already loaded. Hence if you load new meta data from AUTOPILOT_VERSION info that will need to reload. That kind of needs a full reconnect sequence to happen.Also I would say that the meta data files need to be versioned such that:
* The download can be skipped if QGC already has the current version. Means version need to be outside somewhere.
* if there is a newer version, then download and do the reconnect thing.

**Comment by bkueng** [comment]
Closing in favor of #10337

**Pull request closed without merging**